### ISSUE TITLE
Optimize (collapsible) volatile partials markup

### ISSUE THREAD

**Issue by bloatware** [textpattern/textpattern#675]
Due to some recent commits, we need to tighten the construction of future (and current if possible) partials. The good (imo) way is

``` php
        // Partials array, 'something' section.
        'something' => array(
            'mode'     => PARTIAL_VOLATILE,
            'selector' => '#section-something',
            'cb'       => 'article_partial_something',
        ),
...
        // Output 'something' collapsible section or just $partials['something']['html'].
        echo wrapRegion('txp-something-group', $partials['something']['html'], 'txp-something-group-content', 'something_settings', 'article_something');
...
function article_partial_something($rs)
{
    // Construct some $default output from $rs, then let plugins chime in
    ...
    $out = pluggable_ui('article_ui', 'something', $default, $rs);
    // Wrap the output in 'selector', but not collapsible!
    return tag($out, 'div', array('id' => 'section_something'));
}
```

This way partials (including plugins output) will be properly updated on async saves, without changes in collapsible pane states.

**Comment by Bloke** [comment]
Good call. Can't see a downside to this off the top of my head.

**Comment by bloatware** [comment]
I'm not quite happy with extra wrapping divs just for update purposes, but don't see a better backwards compatible way.

**Comment by Bloke** [comment]
No, unless we somehow change `replaceWith()` to something that leaves the wrappers intact. Like, I dunno, `partial.empty().append()`.

**Comment by bloatware** [comment]
Yes, but current `selector` wrappers often contain other partials as well

**Comment by Bloke** [comment]
Ah yes. Drat.

**Comment by philwareham** [comment]
Can you let me know how, if at all, this changes HTML structure? So I can replicate the change in my static mockups.

Thanks.

**Comment by bloatware** [comment]
Hi Phil,

the only change is an extra `<div id="section_something" />` wrapper around the `pluggable_ui()` output. This div will be replaced by a new one on async saves. We can not replace the surrounding `<div class="toggle" />` anymore, since its state (collapsed/expanded) is no longer stored on the server. The whole collapsible section markup is like this

```
<section class="txp-details" id="txp-something-group">
  <h3 class="txp-summary" id="txp-something-group-label">
    <a role="button" href="#txp-something-group-content">something_settings</a>
  </h3>
  <div class="toggle" id="txp-something-group-content" role="group">
    <div id="section_something">Pluggable UI output</div>
  </div>
</section>
```

This is not ideal (an extra div) and could yet change in 4.7, we should discuss it some day.

**Comment by philwareham** [comment]
OK, if it needs an extra `div` then so be it. There are already cases of that in the system (probably for the same reason) such as the additional `div#write-comments` within the comment options section.

**Comment by bloatware** [comment]
That's not that it really needs it, I think we can get rid of it by delegating "toggle" states to `h3` and using CSS sibling rules. But this would mean some CSS modifications (that I'd leave with you) and much testing. If you think it's worth delaying 4.6 release, we can try it in a branch.

**Comment by philwareham** [comment]
A branch would be good - I have some CSS amends still to do to the admin theme so don't mind a bit more.

**Comment by bloatware** [comment]
On second thought, I actually don't see how to avoid extra wrappers, especially when the section contains also static elements.

**Comment by philwareham** [comment]
No worries, as I said an extra div is not a massive issue. As long as we keep a consistent/logical naming system for id/class selectors that is fine.

**Comment by bloatware** [comment]
It looks like everything is patched in 4.7, closing.
